bsc.generics/GenericTests.bs

```diff
@@ -0,0 +1,175 @@
+package GenericTests where
+
+import ListN
+import Vector
+import BuildVector
+
+-- Represents "evidence" for type equality between a and b,
+-- can only be constructed by refl ("reflexive") when a and b are equal.
+-- This could move to a library, I suppose?
+data TypeEq a b = Refl__
+refl :: TypeEq a a
+refl = Refl__
+
+
+data Foo = A (UInt 8) Bool Bar
+         | B (UInt 16)
+         | C
+         | D {a :: (UInt 8); b :: Foo}
+  deriving (Eq, FShow)
+
+struct Bar =
+  foo :: Foo
+  x :: (UInt 8)
+ deriving (Eq, FShow)
+
+data Baz a = Baz a a
+  deriving (Eq, FShow)
+
+struct Qux =
+  x :: a -> a -- Higher rank
+  y :: Int 8
+
+instance Eq Qux where
+  (==) q1 q2 = (q1.x 42) == (q2.x 42) && q1.y == q2.y
+
+instance FShow Qux where
+  fshow q = $format "Qux " q.y
+
+data Kax = K1 (UInt 8) (a -> a) -- Higher rank
+         | K2 (UInt 8)
+
+instance Eq Kax where
+   -- XXX Can't match on higher-rank field due to https://github.com/B-Lang-org/bsc/issues/277
+  (==) (K1 x1 _) (K1 x2 _) = x1 == x2 -- && (y1 42) == (y2 42)
+  (==) (K2 x) (K2 y) = x == y
+  (==) _ _ = False
+
+instance FShow Kax where
+  fshow (K1 x _) = $format "K1 " x
+  fshow (K2 x) = $format "K2 " x
+
+-- Test generic representations
+fooRepr :: (Generic Foo r) => TypeEq r
+  (Meta (MetaData "Foo" "GenericTests" 4)
+   (Either
+    (Meta (MetaConsAnon "A" 0 3)
+     (Meta (MetaField "_1" 0) (Conc (UInt 8)),
+      Meta (MetaField "_2" 1) (Conc Bool),
+      Meta (MetaField "_3" 2) (Conc Bar)))
+    (Either
+     (Meta (MetaConsAnon "B" 1 1)
+      (Meta (MetaField "_1" 0) (Conc (UInt 16))))
+     (Either (Meta (MetaConsAnon "C" 2 0) ())
+      (Meta (MetaConsNamed "D" 3 2)
+       (Meta (MetaField "a" 0) (Conc (UInt 8)),
+        Meta (MetaField "b" 1) (Conc Foo)))))))
+fooRepr = refl
+
+barRepr :: (Generic Bar r) => TypeEq r
+  (Meta (MetaData "Bar" "GenericTests" 1)
+   (Meta (MetaConsNamed "Bar" 0 2)
+    (Meta (MetaField "foo" 0) (Conc Foo),
+     Meta (MetaField "x" 1) (Conc (UInt 8)))))
+barRepr = refl
+
+bazRepr :: (Generic (Baz a) r) => TypeEq r
+  (Meta (MetaData "Baz" "GenericTests" 1)
+   (Meta (MetaConsAnon "Baz" 0 2)
+    (Meta (MetaField "_1" 0) (Conc a),
+     Meta (MetaField "_2" 1) (Conc a))))
+bazRepr = refl
+
+bazFooRepr :: (Generic (Baz Foo) r) => TypeEq r
+  (Meta (MetaData "Baz" "GenericTests" 1)
+   (Meta (MetaConsAnon "Baz" 0 2)
+    (Meta (MetaField "_1" 0) (Conc Foo),
+     Meta (MetaField "_2" 1) (Conc Foo))))
+bazFooRepr = refl
+
+{- Can't actually test this, due to the presence of a generated struct Qux_x
```

Comment:
Yes, because this gets created in the deriving phase after initial type checking has already happened.  I'm not really sure if there is a reasonable fix for this.

Reply:
Is the problem that `Qux_x` is marked as an internal identifier of some sort? Maybe we can just not do that?

Reply:
No, the problem is this `Qux_x` literally doesn't exist in the symbol table before the deriving step, which causes a compilation failure before we even know that such a wrapper struct needs to be generated.

Reply:
Oh, right. Kind inference would fail for Qux_x. I don't think there's an easy way around that especially since you need some symbol table for deriving.

Reply:
I was confused, because the `typecheck` stage occurs after `deriving`, so it's not the typecheck stage that is complaining.

The problem is that `Qux_x` appears in the explicit type for a top-level definition.  The error occurs when constructing a symbol table of all available top-level definitions.  (So I suspect that `Qux_x` is visible to users in the body of defs, just not in their top-level types.)

If we want the new types to be predictably created and visible to users (even in top-level types), we would need to insert that creation into the building of the symbol table (rather than in the deriving stage).

I thought it might be as easy as changing the definition of `MakeSymTab.getTI` on structs to insert the additional types theres (when the struct is an interface with poly methods).  However, in `mkSymTab`, we get all the names (`dids`) and check for undefined names (`uids`) before we call `mkTypeSyms` (that calls `getTI`).  Perhaps we can just do a second filter of the `uids` after constructing the type info?

Reply:
No, wait, maybe you can just insert the type prior to making the symbol table.  I don't think you need a symbol table at all -- you just look through the package type declarations for interfaces and I think you can identify the polymorphic ones just from knowing the type variables of the interface and the variables of the method, which are evident from syntax.  So you could insert a stage after `opparse` and before the first symbol table construction.

Reply:
Adding an extra stage in the compiler still seems like a lot of added complexity for something that is already somewhat of a corner case.  

While these types can show up in error messages, I can't imagine a reason (besides this specific test case) why someone would need to reference one of these generated structs in user code.  For example when implementing the class `MyClass` generically, if faced with the error message
```
No instance for MyClass' (ConcPoly Qux_x)
```
the fixes would be to either
* write an explicit instance for `MyClass Qux`;
* write an instance of `MyClass' (ConcPoly a)`, which may be an actual implementation (e.g. ignore the field in some manner), or (in the future) raise a better error message using a `TypeError` context.

Neither resolution requires actually referencing the `Qux_x` type from user code (if you are directly referencing `Qux_x`, then you are special-casing on a particular type and nothing is gained from doing it through the generic representation) so I don't see how being able to name the type in user code would help the situation. Additionally having generated types that can't be referenced from user code but do show up in other ways isn't unprecedented - for example the `SDataCon` generated structs.  

An alternative would be to try and hide these types from the user as much as possible and somehow exclude them from the Bluetcl output.  But even still I don't see a good way of avoiding these generated types appearing in error messages, so this alternative seems fruitless.  

So given these considerations would you be alright with leaving things as they currently stand, rather than spending more time on this issue?

Reply:
I am okay with not adding an additional stage.  However, I do want to investigate whether we should be hiding the new type from the bluetcl output.

The output of the `polyfield.tcl` test has changed because the new types appear in the output of a call to `defs type`.  And if I look at the implementation of this command in `bluetcl.hs`, I see that the case arm of `tclDefs` for `type` has a filter to remove IDs that we want to hide:
```
let ts = filter (not . hideId) (concat tss)
```
where `hideId` is defined later in the file, to ignore a number of different IDs including those marked as "internal".

However, I notice that the construction of the `SDataCon` internal type does not mark the type as internal (in neither the Classic nor BSV parser).  And I confirmed, by writing an example, that `defs type` does print the internally generated type.

I would think that `SDataCon` should be marked internal (with `Id.setInternal`), however I would want to confirm that it doesn't break anything, and that's probably beyond this PR.  For one, it's probably only safe to be hiding the internal struct if the display of the parent type is also hiding the presence of the substruct (and instead providing a source view of the parent type, that shows it having the fields that are actually in the substruct).  In contract, types like `Qux_x` don't have any useful information that we need to worry about hiding.  So I don't think there will be any gotchas to hiding them.

If you want to leave things as they are (not hiding the new types), that's fine.  But could you try applying `setInternal` to the ID in the type declaration and seeing if that breaks anything? That would be my preference.

Alternatively, the `hideId` also checks for IDs that start with `__` (a convention that's use in some places in the compiler) and ignores those as well.  I had also expressed concern about whether these new names might accidentally conflict with user names; if you called the type `__Qux_x`, then you'd both reduce the possibility of clash and get hiding for free.

Reply:
It may also be useful to compare how GenWrap creates the new type for the bitified interface (of the synthesized module that it wraps).  I believe that `GenWrapUtils.ifcIdRename` is what is used to created the new name, in most cases.  It uses `addInternalProp`.  I think this name is just the original type with its arguments appended (so for example: `RegBool`) (and possibly with other annotations appended).  But I see that in GenWrap there is also a line that creates a new ID with `setInternal` and `enumId` (that sets the "bad" property as well).  This is creating a name of the form `_<ifcname>_<pos>` (using the position to help make it unique).  I don't offhand recall what situations this is used in.

One difference with `Qux_x` is that it appears in error messages, so you may not want it to be very mangled.  So perhaps just giving it the "internal" property is sufficient.

Reply:
I thought that I had fixed this by setting the "internal" property - but apparently these generated structs no longer appearing in the Bluetcl output was one of the consequences of removing the deriving exemption for `Id__` (the existence of an instance `Generic (Id__ a) (ConcPrim a)` meant that `Generic` instances were no longer being generated for types outside of the `Prelude`.   

So apparently the internal property is not having an effect on these types being included in Bluetcl, or somehow isn't being propagated properly.  Any suggestions on where to look for where this might be going wrong?

I did also try instead adding the `__` prefix on the names of these structs.  However it seems that Bluetcl only checks for this prefix on names that are defined in the Prelude.

Reply:
Does the bluetcl command only report types that are exported? Sometimes a type gets automatically added to the export list if it's used in the type of something that is exported.  I don't know if instances count towards that.  (And what types are mentioned in top-level prototypes can be reduced by the CtxReduce stage, if it can show that some contexts can be removed.)  Those are some thoughts off the top my head, I can have a closer look later.

Reply:
I notice things:

(1) The bluetcl command where we want to do the filtering is `defs`.  This command takes all the top-level defs and filters specific ones to show.  We're executing `defs type PkgName` which shows all the top-level type declarations in the package.  It filters the IDs that need hiding.  The ID that needs the property, then, is the ID in the top-level def.  That is the `CStruct` being created in `mkGenericRepWrap` (in `Deriving`):
```
mkGenericRepWrap r packageid pos i ty_vars fty@(CQType _ ty) =
  [Right [Cstruct True SStruct (IdK i) vs fields []],
  ...
```
At one time you had code to set the "internal" property on that ID, but you seem to have removed it.  (There are two other places in `Deriving` where you appear to be attempting to set the "internal" property.  I think you can remove those?)

(2) For the two places where your PR is still attempting to set the "internal" property (in `Deriving`), you've imported `addInternalProp` from `GenWrapUtils`.  That function isn't actually adding the "internal" property!  It may have done so at one time (and thus the name), but it is currently tagging the ID as being a "generated ifc".  That function should probably be renamed `addGenProp` or something (noting that there's `isGenId` defined next to it, for checking when an Id has that property).  For `Qux_x` types, I think you should instead just use `setInternal` from the `Id` package.  (Note that `GenWrap` does also use `setInternal` in one place where it creates a type declaration.  That would be another reason to rename the uses of `addInternalProp`, to distinguish them.  Apologies if I added to this confusion!)  (Also, I would not import something from GenWrap into Deriving, anyway.  If it were a function worth being common to the two, then it should be moved to a more common location.)

If I add `setInternal` to `mkGenericRepWrap` (and remove the uses of `addInternalProp`), then the bluetcl test goes back to the old behavior (of not displaying the `Qux_x` types.)

Reply:
Fixed.